csi/src/node/iscsiutil.rs

```diff
@@ -79,67 +84,95 @@ fn attach_disk(
-    ];
-    trace!("iscsiadm {:?}", args_login);
-    // login to iscsi target
-    let output = Command::new(&iscsiadm)
-        .args(&args_login)
-        .output()
-        .expect("Failed iscsiadm login");
-    if !output.status.success() {
-        let args_login_del = [
-            "-m", "node", "-p", &tp, "-T", &iqn, "-I", "default", "-o",
-            "delete",
+    // Do not attempt to create a session if one exists.
+    if !have_session {
+        let target = format!("{},{} {}", tp, lun, iqn);
+
+        let args_discovery =
+            ["-m", "discovery", "-t", "st", "-p", &tp, "-I", "default"];
+        trace!("iscsiadm {:?}", &args_discovery);
+        let output = Command::new(&iscsiadm)
+            .args(&args_discovery)
+            .output()
+            .expect("Failed iscsiadm discovery");
+        if !output.status.success() {
+            return Err(Error::from(CSIError::Iscsiadm {
+                error: String::from_utf8(output.stderr).unwrap(),
+            }));
+        }
+
+        // Check that the output from the iscsiadm discover command lists
+        // the iscsi target we need to login.
+        // If not fail.
+        let op = String::from_utf8(output.stdout).unwrap();
+        let haystack: Vec<&str> = op.split('\n').collect();
+        if !haystack.iter().any(|&s| s == target.as_str()) {
+            trace!("After discovery no record for {}", target);
+            return Err(Error::from(CSIError::Iscsiadm {
+                error: format!("No record for {}", target),
+            }));
+        }
+
+        let args_login = [
+            "-m", "node", "-p", &tp, "-T", &iqn, "-I", "default", "--login",
         ];
-        // delete the node record from the database.
-        Command::new(&iscsiadm).args(&args_login_del);
-        return Err(Error::from(CSIError::Iscsiadm {
-            error: String::from_utf8(output.stderr).unwrap(),
-        }));
+        trace!("iscsiadm {:?}", args_login);
+        // login to iscsi target
+        let output = Command::new(&iscsiadm)
+            .args(&args_login)
+            .output()
+            .expect("Failed iscsiadm login");
+        if !output.status.success() {
+            let args_login_del = [
+                "-m", "node", "-p", &tp, "-T", &iqn, "-I", "default", "-o",
+                "delete",
+            ];
+            // delete the node record from the database.
+            Command::new(&iscsiadm).args(&args_login_del);
+            return Err(Error::from(CSIError::Iscsiadm {
+                error: String::from_utf8(output.stderr).unwrap(),
+            }));
+        }
     }
 
-    if wait_for_path_to_exist(device_path.as_str(), 10) {
+    let millis: u64 = 1000;
+    if wait_for_path_to_exist(device_path.as_str(), 10, millis) {
         trace!("{} path exists!", device_path)
     } else {
-        trace!("{} path does not exist after 10s!", device_path);
+        trace!(
```

Comment:
So let us assume we timeout, do we clean up the session and the discovery entry?

Reply:
No, that would be detrimental, I think. Because if the connection didn't succeed within an arbitrary time limit, it unlikely to succeed if everything is torn down and setup again with the same time limit.
The objective is to allow the connection process to proceed in parallel so that the next time attach is called it checks on the session that may have been initiated previously, in effect allowing the iscsi connection to proceed unimpeded across that time.